### DISCUSSION TITLE
Resume Workflow

### DISCUSSION THREAD

**Discussion by FarahmandMhmd** [elsa-workflows/elsa-core#1214] [Q&A]
Hi
I want to run very simple workflow to understand the basics and then extend it.

1- User send json { id: 1 , name: "Elsa } to **UserController** 
2- UserController trigger **SignalReceived** 
3-  signalReceived trigger OnExecute() method in **RegisterActivity** and suspend.

use postman to post data to controller
![aaaa](https://user-images.githubusercontent.com/80684177/124470965-8a0d2800-ddb1-11eb-97ab-f0d5aa6cc9b9.jpg)

**UserController**
`

    [Route("api/[controller]/[action]")]
    [ApiController]
    public class UserController : ControllerBase
    {

        private readonly ISignaler _signaler;

        public UserController(ISignaler signaler)
        {
            _signaler = signaler;
        }

        [HttpPost]
        public async Task<string> Save(UserModel user)
        {
            var startedWorkflows = await _signaler.TriggerSignalAsync("UserWorkflow", cancellationToken: default, correlationId: user.Id.ToString());

            string message = "data received";

            return message;
        }
    }

`

**//RegisterActivity**

`

    public class RegisterActivity : Elsa.Services.Activity
    {
        protected override IActivityExecutionResult OnExecute()
        {
            return Suspend();
        }

        protected override IActivityExecutionResult OnResume(ActivityExecutionContext context)
        {
            // Read received input.
            var receivedInput = context.GetInput<string>();

            // Store received input into activity output.
            context.Output = receivedInput;

            // Instruct workflow runner that we're done.
            return Done();
        }
    }

`

Everything work fine!
Now problem is how to resume workflow after Admin send json { userId: 1 , confirm : true } to **AdminController**

`

    public class AdminController : ControllerBase
    {
        private readonly ISignaler _signaler;

        public AdminController(ISignaler signaler)
        {
            _signaler = signaler;
        }

        [HttpPost]
        public async Task<string> Confirm(AdminModel admin)
        {
           // need code to trigger and resume  suspended workflow
            string message = "Confirmed";
            return message;
        }
    }

`

**Comment by sfmskywalker** [comment]
In this example, there's no need to manually start the workflow first and then resume, because you are using the **Signal Received** activity as a starting activity, which will act as a workflow trigger.

All you need to do trigger the workflow in your screenshot is this:

```csharp
var startedWorkflows = await _signaler.TriggerSignalAsync("RegisterWorkflow", cancellationToken: default, correlationId: user.Id.ToString());
```

If you want an approval workflow where the admin user approves the registration, you need to add another **Signal Received** activity.

For example, you might have the following:

1. Signal Received activity with signal "Register".
2. Signal Received activity with signal "Confirm".
3. RegisterActivity (your custom activity) that actually performs the registration.

**Reply by FarahmandMhmd** [reply]
I apologize for asking stupid questions!
Although i read the full documentation on Github , I still don't understand how Elsa works.
it's been almost a week since I started learning Elsa workflow.

![Untitled](https://user-images.githubusercontent.com/80684177/124708604-bd65c900-df0f-11eb-8715-833f9bf766d0.png)

Now when User send registration request , Signal Received ("Register") passed user Id (for example id=5) as  correlationId to OnExecute() method in registerActivity and suspended. it is also saved in Elsa database . 

Now when admin send confirm request , i want to pass user Id (id=5) through Signal Received ("Confirm")  to OnResume() method to resume previous suspended correlationId( id = 5) that user sent.

But Signal Received ("Confirm") can not trigger registerActivity when i send id ( id=5) that has already been sent and suspended.
it works when admin send new id ( for example id = 6) , but it will trigger OnExecute() method again  !  .

in other words both Signal  'Confirm' and 'Register' work the same way.

**Reply by jdevillard** [reply]
Hello, don't do a ParallelBranch but stay linear in your workflow.

as explain on discord, you don't have to suspend in the RegisterActivity.
let's assume that you have the following step in your workflow : 

1. User Want To Register and Signal UserWorkflow
2. The Workflow is Starting the registration  (using StartRegisterActivity that will do something but don't need to suspend())
3. The Workflow is waiting for the Admin Approval using a Admin Signal
4. When the Admin Signal is received, the Workflow finish the Registration  (using CompleteRegistrationAfterAdminActivity that will do something but don't need to suspend())
5. The Workflow finished.

```
    public class RegisterWorkflow : IWorkflow
    {
        public void Build(IWorkflowBuilder builder)
        {
            builder
                .SignalReceived("UserWorkflow")
                .Then<StartRegisterActivity>()
                .WriteLine((ctx) => $"Signal received! -- Now Registering Activity and Waiting for admin signal with Instance Id {ctx.WorkflowInstance.Id}")
                .SignalReceived("AdminSignal")
                .WriteLine(() => $"Signal From Admin Received")
                .Then<CompleteRegistrationAfterAdminActivity>()
                .WriteLine(() => $"End of Worflow")
                ;
        }
    }

 public class StartRegisterActivity : Elsa.Services.Activity
    {
        protected override IActivityExecutionResult OnExecute()
        {
            Console.WriteLine("doing something to start the registration process and wait for validation");
            return Done();

        }
    }
    public class CompleteRegistrationAfterAdminActivity : Elsa.Services.Activity
    {
        protected override IActivityExecutionResult OnExecute()
        {
            Console.WriteLine("doing something to validate the registration after Admin Signal");
            return Done();

        }
    }
```

The workflow can be tested like this in a Console App : 

```
           Console.WriteLine("Send UserWorkflow Signal");
            var signaler = services.GetRequiredService<ISignaler>();
            await signaler.TriggerSignalAsync("UserWorkflow");

            // Copy the Instance Id printed by the workflow
            var instanceId = Console.ReadLine();

            await signaler.TriggerSignalAsync("AdminSignal",null, instanceId, null);

            Console.ReadLine();
```

With the explanation : 
```
           Console.WriteLine("Send UserWorkflow Signal");
            var signaler = services.GetRequiredService<ISignaler>();
            await signaler.TriggerSignalAsync("UserWorkflow");

```
is what you already did in your UserController 

and 
```
 await signaler.TriggerSignalAsync("AdminSignal",null/* content if needed */, instanceId, null);
```
what you have to do in your admin controller.
So keep in mind that you have to get the good instanceId to resume the good workflow.

HTH

**Reply by FarahmandMhmd** [reply]
Thanks , please check below workflow

**Comment by FarahmandMhmd** [comment]
Hello jdevillard ,
I implemented the workflow besed on your codes but admin can't trigger and resume the workflow!

![Untitled](https://user-images.githubusercontent.com/80684177/125153190-4aef2600-e167-11eb-8abe-d6e2ee031da3.jpg)

i sent json { "Id": "1",   "Name": "Elsa"} via postman to user controller and trigger **StartRegisterActivity** .

 var startWorkflow = await _signaler.TriggerSignalAsync("User", cancellationToken: default, correlationId: user.Id.ToString());

![Untitled](https://user-images.githubusercontent.com/80684177/125153321-20519d00-e168-11eb-97a6-a5dc8a017846.png)

the correlationId has been successfully saved in Elsa database as suspended.

Now when i send Json { "userId": "1",   "confirm": "true"} to admin controller , 
var completeWorkflow = await _signaler.TriggerSignalAsync("Admin", cancellationToken: default, correlationId: admin.UserId.ToString());

it can not trigger onExecute() function in **CompleteRegisterActivity** !!!

**Reply by sfmskywalker** [reply]
Could you share your project so we can take a look? something is missing, but it's hard to say what. From what I've seen so far, it should work.

**Reply by FarahmandMhmd** [reply]
Hello sfmskywalker , 
sure! 
you can find it below
https://github.com/FarahmandMhmd/Workflow

**Reply by jdevillard** [reply]
Can you please share the implementation of your two register activities ?

**Reply by FarahmandMhmd** [reply]
project source code is in below Github address.
https://github.com/FarahmandMhmd/Workflow
check activity folder

**Reply by jdevillard** [reply]
@FarahmandMhmd , 

as explain in the previous example 
```
await signaler.TriggerSignalAsync("AdminSignal",null, instanceId, null);
```

it seems that the TriggerSignalAsync use the instanceId and not the correlationId when you need to resume an instance.

so if you get the instance Id in usercontroller and use the instanceid in the adminController, it works (tested on my side with your  code)



_The correlation id parameter seems to be use to set the CorrelationId for the instance but not to resume._
@sfmskywalker  can you confirm?

**Reply by FarahmandMhmd** [reply]
nice! it works fine.
thank you **jdevillard**
If this is the correct way to resume workflow , how can i get the instance ID? 
form local Elsa database or Elsa API ?
I'll be waiting for **@sfmskywalker** confirmation.

**Reply by jdevillard** [reply]
To add some element on the correlationId, 

The CorrelationId will ensure that _you don't start more than ONE workflow for you user, so you can continue to specify the correlationId to start the Workflow_. 

In you UserController, you can get the WorkflowInstanceId like this (just an example) : 
```
var startWorkflow = await _signaler.TriggerSignalAsync("User", cancellationToken: default, correlationId: user.Id.ToString());
string message = $"data received - {startWorkflow.ToArray()[0].WorkflowInstanceId}";
```         

so you can save this Workflow Instance Id to allow the resuming.

Regarding the CorrelationId, here , if you call twice the api/User/Save api : 

1. the first call to TriggerSignalAsync will return an Enumerable with the instance of the newly created workflow, to you can get it. 
2. the second call (because the Workflow is already created and in suspend mode with the same correlationid) will return null because no new Workflow will be created.

**Reply by FarahmandMhmd** [reply]
Perfect!
Thank you for taking time to answer in detail.

**Reply by sfmskywalker** [reply]
> The correlation id parameter seems to be use to set the CorrelationId for the instance but not to resume.

The correlation ID is used differently depending on what the `TriggerSignalAsync` will find in terms of startable workflows and resumable workflows.

For instance, if there is a **workflow blueprint** that starts with the specified signal, then three things will happen:

1. A new workflow instance is created.
2. The new workflow instance is correlated with the provided correlation ID.
3. The new workflow instance will immediately execute the next activity (it does not get blocked on the **Signal Received** activity because it was used as the starting activity and therefore acts as a workflow trigger and not as a blocking activity).

If there is a workflow instance that is blocked on the specified signal, then 1 thing will happen:

1. The existing workflow instance matching the provided correlation ID will be resumed.


> The CorrelationId will ensure that you don't start more than ONE workflow for you user, so you can continue to specify the correlationId to start the Workflow.

Exactly.

**Reply by FarahmandMhmd** [reply]
lets consider two separate applications.
a website that user can register for shopping (app1)
a web app (admin panel) for confirm or reject registration(app 2)

**user Id** is unique for CorrelationId 

based on @jdevillard explanation,  one way to resume workflow after user registration is that we get current WorkflowInstanceId with below code: 

       var startWorkflow = await _signaler.TriggerSignalAsync("User", cancellationToken: default, correlationId: user.Id.ToString());
        var instanceId = startWorkflow.ToArray()[0].WorkflowInstanceId;

and then save WorkflowInstanceId to database.
admin will be able to trigger and resume workflow with this saved WorkflowInstanceId .

If we don't want to get and save WorkflowInstanceId  for each registration request , is there another way to resume workflow?

In other words , without saving WorkflowInstanceId in database , we can get WorkflowInstanceId  in app1 , but in app2 , admin only has access to user Id (CorrelationId) due to the separation of app1 and 2.

Have you considered a method in Elsa API to get WorkflowInstanceId with CorrelationId?

**Reply by sfmskywalker** [reply]
There’s a bug in Elsa 2 (or just 2.1, not sure) that requires workflow instance ID. But I’m pushing a fix so that you don’t need a workflow instance ID, just the correlation ID (User ID in your example).

Nonetheless, having an API to get workflow instance ID(s) by correlation ID (and optionally workflow definition ID) might be good to have all the same.

**Reply by FarahmandMhmd** [reply]
I want to use if/else for admin approval in my example but could not find detailed explanation of the If/Else control flow in Elsa 2 .

admin send Json { userId: 1 , confirm : true } to admin controller and trigger SignalReceived "Admin".

var adminApproval = await _signaler.TriggerSignalAsync("Admin",  input: admin.confirm, workflowInstanceId: admin.UserId.ToString());

In defenition of **TriggerSignalAsync** , there is input object.
I want to store admin approval as true or false in input object and send it to if/else control flow to check if true then continue and complete registration and if false only send email to user .

Is is possible to connect signalReceived admin to If/else control flow directly and check input object in if/else?
or need activity (AdminApprovalActivity) between them to outcome true or false?

What is the input of If/else control flow and how it calculates the output via javascript?
can you explain it with an example that shows javascript syntax for checking true or false in this case?

Which one is better?

![New Project (1)](https://user-images.githubusercontent.com/80684177/125569026-1e41e169-013b-400e-8906-981be1626b41.jpg)

**Reply by jdevillard** [reply]
Hello, 

the Ouput of the SignalReceived Activity will be the Input Object you passed in the TriggerSignalAsync.

If you take a look at the Workflow Instance, you will see the value : 
![image](https://user-images.githubusercontent.com/3254429/125955185-f24a023f-1ca9-42d6-877e-700e1c1b2ae2.png)

After, to access your Trigger Object, you can use javascript expression to get the input which is passed to the next Activity.
In you case, if you want to set the IF/Else just after , you just have to use in javascript : 
```
input.Confirm
```

input is the object passed to the activity and Confirm is one of the property (the object is read as a Json).
This will be done like the capture below :

![image](https://user-images.githubusercontent.com/3254429/125955044-42216f3c-e6ff-44ce-bee9-d124f0b3451b.png)


If you want to use the object for more than the next activity, you have to first save the value as a variable and then you can get it from javascript OR you can use the latest version of Elsa :) . 

It seems that you used the 2.0, prior to 2.1  the definition of activity properties is a little bit modified and there is now a notion of output.

For example, after setting a name to your activity :
![image](https://user-images.githubusercontent.com/3254429/125956154-c644c09f-40a5-49dd-bcec-b1b77256e690.png)

you will be able to get the Output object Signal Activity (the input object of TriggerSignalAsync) for every activities you want using the following expression (and with intellisense)

![image](https://user-images.githubusercontent.com/3254429/125956497-9600501e-baf0-4251-bd2c-dbe8ac2818e8.png)


and to access the Confirm Property of the object  : 
```
activities.AdminReceivedActivity.Output().Confirm
```

**Reply by FarahmandMhmd** [reply]
I made a mistake in writing javascript expression.
if (input.Confirm) { } 
always returning false   :-)
Thank you so much for explaining in full detail.
I'm sure this example can help others who are new in Elsa.

**Comment by FarahmandMhmd** [comment]
Hi @jdevillard , as you explained no need to suspend workflow in this example.

What do you think if add dropdown to workflow to choose Admin?

![2](https://user-images.githubusercontent.com/80684177/126140092-ad3b3532-47d8-4ae8-99cd-0f142b01ec92.png)
![1](https://user-images.githubusercontent.com/80684177/126140108-ee956519-5948-4cf4-8fd3-d6dcc209d72b.png)

1- user send { id: 1 , name : Elsa } to SignalReceived **User**
2- admin send { userId: 1 , confirm: true , adminId : 2} to SignalReceived **Admin**
3-  custom activity **Approval**  check  if (confirm = true & adminId = selected dropdown value) , then resume workflow.

still no need to suspend?!
if not , How to resume suspended workflow?! (my first question in this Q/A)  :))))))

**Reply by FarahmandMhmd** [reply]
I read your message in Discord.

Let's assume this scenario:

1- User send registration request  { Id: 1, username: Elsa}
2- The Workflow is starting the registration.
3- The Workflow is waiting for the Admin Approval.
4- Admin send confirmation { Id: 1 , confirm : true , adminId: 2} 
(each registration request assigned to specific admin(for example admin-1 , admin-2 , admin-3) based on registration type)
5- workflow check if (adminId = selected dropdown value & confirm = true) then workflow continues and finish the registration.

Take note i want to have dynamic workflow that when i select admin-2 in dropdown , workflow continues only with registration request with adminId = 2.


Another example that shows what i mean  is leave request.
company-1 : for leave request , only need technical manager approval.
company-2 : for leave request , only need secretary approval.
company-3 : for leave request , only need CEO approval.

now if we sell our HRM software to them , without workflow we have to change code for each company for leave request .
I want to implement workflow that customer can choose(in workflow) who is responsible for approving the leave request without coding.

Is it possible to create this workflow with standard activity and Kiss principle?

**Reply by jdevillard** [reply]
Hello, 

sorry for the delay, it's summer ;)

what you have to understand about suspending a workflow is that some activities are used to suspend workflow AND resume the workflow instance on a specific trigger/bookmark/event.

The purpose of **Signal Received** activity is to supend the workflow instance and wait for an Event. So When you are on the ApprovalActivity, the workflow engine has already : 

1. receive an event
2. resume the workflow
3. instanciate the ApprovalActivity activity

You always need to suspend if you need to wait for an external event. In your case, the event is an admin approval calling a web API so you Workflow need to suspend to wait for this event.

Now you problem is that you have an Admin Event with following properties : 

- adminId
- confirm

and the Workflow is resume because you receive an AdminEvent AND NOT if you receive an adminEvent WITH confirm==True AND adminId == selected.

Regarding this, you have multiple options : 

1. resume the Workflow when you receive an admin event and then check for Confirm and AdminId Properties like this :

set a confirmation variable at the beginning of the Workflow : 
![image](https://user-images.githubusercontent.com/3254429/127213704-32ad8668-8495-4d29-8768-5a3d137c8a81.png)

then use a while loop the check if the event received by the signal is good regarding your configuration : 
![image](https://user-images.githubusercontent.com/3254429/127213751-9539203c-bf52-487d-b09b-a4f27102ee0e.png)

in detailled , in the while you check you confirmation variable
![image](https://user-images.githubusercontent.com/3254429/127213953-a054b23f-52cb-42af-a73e-a2162158dc4b.png)

in the IF/Then you check if the model transmitted by the signal correspond to your need  (ie: confirm == true And AdminId == selected)
![image](https://user-images.githubusercontent.com/3254429/127214021-3b862429-9fb4-4bd9-b327-ab062aeab62e.png)

if not correct, you continue to wait for another event (with the same CorrelationId or Workflow Instance Id ) , if OK you change the variable "confirmation" and you exit the loop and you can approve you process.

as you can see here , you don't need to create a new specific activity , you only need to use the built-in activities (the COMPARISON expression is less well than a dropdown but require less dev... :). If you want to use a custom activity, you only need to pass the Admin Model to you activity and then check the value regarding the dropdown and fill the good outcome. (But this will just remove the IF/THEN activity, you will always need to loop while the confirmation is good)


2/ You can create a custom activity that will suspend the workflow, and wait for specific signal using new bookmark etc... but you will need to create specific code, and I don't think it will be a good idea for your need.

Hope this Help
